### PR TITLE
Add support for accessing Ceph RBD-based volumes with the rbd-nbd client

### LINKED ISSUE #32266: Add support for accessing Ceph RBD with the rbd-nbd client
## Objective

Add support for the recently emerged `rbd-nbd` feature-rich client to improve support for Ceph RBD in Kubernetes.
## Background

Kubernetes already offers support for Ceph RBD based on the `krbd` kernel client. Unfortunately, due to obvious reasons, `krbd` can't use the `librbd` user-space space library that gets most of the development focus. This caused a feature gap when compared to VM-based environments.

At the moment `krbd` lacks exclusive locking infrastructure, image management facilities and multi-DC support (remote backups!). The detailed list can be obtained by comparing the definitions of `RBD_FEATURES_ALL` macro in [`librbd`](https://github.com/ceph/ceph/blob/master/src/include/rbd/features.h#L25) and [kernel](http://lxr.linux.no/#linux+v4.7.2/drivers/block/rbd.c#L121). While there are efforts on partially closing the gap (patches waiting for merge), `librbd` continuously gets new items like [consistency groups](http://tracker.ceph.com/projects/ceph/wiki/consistency_groups) or [fencing improvements](https://github.com/ceph/ceph/pull/9592).

Beside the feature gap `krbd` exhibits additional drawback. Being entirely kernel space impacts fault-tolerance as any kernel panic affects a whole node -- not only a single Pod using RBD storage. According to the [`rbd-nbd` pull request](https://github.com/ceph/ceph/pull/6657) the reliability, especially on non-x86 platforms, was one of the driving factors behind the new client.

Those issues can be addressed by employing `rbd-nbd` -- a thin adapter ([858 lines of code in Jewel](https://github.com/ceph/ceph/blob/v10.2.2/src/tools/rbd_nbd/rbd-nbd.cc#L858)) between NBD subsystem of Linux kernel and `librbd`.

The primary concern about `rbd-nbd` was performance. The authors claimed it's _comparable_ to krbd. Together with Intel we verified those claims. Tests on typical HW where Ceph is deployed (SSDs for journals, HDDs as main storage) showed `rbd-nbd` delivers 70-80% of `krbd` performance. Though, we don't consider the raw performance as a deciding factor. Even fastest RBD client wouldn't be a solution-of-choice for performance critical use cases like RDBMS hosting.
## Proposal

We would like to propose incorporating support for `rbd-nbd`. The integration can be accomplished in multiple ways.
1. Extending the current in-tree plugin for RBD to use `krbd` or `rbd-nbd` depending on a configuration parameter. This would assure the excellent commonality. On the other hand we see differences where this might become an issue.  
   The _krbd_-based plugin employs the _advisory locking_ mechanism to provide fencing at the cost of HA -- in case of node crash a manual intervention is required. However, serving images marked to use _exclusive locking_ feature would result in employing the same underlying infrastructure twice. At the moment this isn’t an issue because `krbd` simply doesn’t support exclusive lock. However, when it finally gets it, RBD images created with Ceph’s default capabilities set will be inaccessible in read-write mode. This was the reason why a [similar solution](https://lists.gnu.org/archive/html/qemu-devel/2016-04/msg02422.html) has been [refused](https://lists.gnu.org/archive/html/qemu-devel/2016-04/msg04131.html) in QEMU.  
   Another problem we see are additional dependencies. After fencing capability is integrated directly in `krbd`, the dependency on user-space `/usr/bin/rbd` CLI utility can be removed. The helper is not required for mapping RBD images as the kernel interface may be used directly by the plugin.
2. Dedicated in-tree plugin. Offers greatest flexibility in adopting new features (like dynamic provisioning of Persistent Volumes).
3. Dedicated FlexVolume plugin. Simple but limited. Bringing support for e.g. dynamic provisioning would require extensive changes in FlexVolume’s interfaces.

At the moment the option # 2 looks most promising. Though we stay entirely open for the discussion.
### Performance

`rbd-nbd` efficiency suffers from extra context switches and extensive memory copying between kernel and user-space boundaries. Although we can do nothing with the first factor, the second can be improved for large blocks transfers. Linux kernel has facilities for zero-copy page-aligned memory transfers between kernel and application via the combination of `splice` and `vmsplice` system calls.
### Fencing

Actually Mirantis [is working](https://github.com/ceph/ceph/pull/9592) on bringing fencing to `librbd` on the top of exclusive lock infrastructure. This would allow to move from the flawed _advisory lock_ mechanism that is currently used by rbd storage plugin.
### Dynamic provisioning

Currently RBD images must be created ahead by a cluster administrator. We see possibility to implement the dynamic provisioning for RBD in the future.

### PR DESCRIPTION
**Pull request by rzarzynski**
**What this PR does / why we need it**:
This PR improves Ceph RBD support by allowing the `pkg/volume/rbd` to use an additional type of RBD client: `rbd-nbd`. As the client bases on the common and broadly adopted `librbd` library, the feature gap between containerized and VM-based environments can be closed. For more details please refer to the issue #32266.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*:
fixes #32266

**Special notes for your reviewer**:
* The plugin got a very thin abstraction layer over mapping and unmapping RBD volumes.
* At the moment two implementations of the AL (backends) exist: `krbd`-based (the default one) and  `rbd-nbd`-based (new). A third implementation covering TCMU can be easily added in the future.
* As the plugin handles also volume provisioning, sharing the code across backends might be profitable. Anyway, moving to a separated plugin called eg. `rbd-userspace` shouldn't an issue at all.

**Release note**:

```
Ceph RBD volumes can be accessed also using the rbd-nbd user space client.
```

### CODE REVIEW

**Comment by rootfs** [comment]
@k8s-bot ok to test

**Comment by rootfs** [comment]
A quick comment. I'd like to see a fallback path for rbd-nbd to rbd, in case rbd-nbd is absent.

**Comment by rzarzynski** [comment]
> A quick comment. I'd like to see a fallback path for rbd-nbd to rbd, in case rbd-nbd is absent.

@rootfs: yeah, having the fallback there sounds interesting. I will take care of that. Thanks!

**Comment by rootfs** [comment]
@rzarzynski please run `hack/update-all.sh` too

**Comment by rzarzynski** [comment]
I was struggling with the code auto-generation for a while but I missed `hack/update-all.sh`. Thanks a lot for pointing it out.

File: pkg/volume/rbd/rbd_util.go [hunk 1]
```diff
@@ -224,94 +82,72 @@ func (util *RBDUtil) loadRBD(mounter *rbdMounter, mnt string) error {
...
-			return errors.New("Could not map image: Timeout after 10s")
-		}
+	// map a block device
+	mapper, err := createDiskMapper(b)
+	if err != nil {
+		return fmt.Errorf("rbd: cannot create diskMapper: %v", err)
 	}
+	devicePath, err := mapper.MapDisk(b)
+	if err != nil {
+		return fmt.Errorf("rbd: cannot map block device")
```
Comment (zhsj) [inline pkg/volume/rbd/rbd_util.go]: Could you return err detail here, when I test this pr, its hard for me to debug the error.

**Comment by kfox1111** [comment]
This still looks useful.

File: pkg/volume/rbd/rbd_util.go (hunk 1 above)
Comment (elaron) [reply pkg/volume/rbd/rbd_util.go]: May I working on this patch? This patch is very important for ceph supporting persistent volume.

Comment (dims) [reply pkg/volume/rbd/rbd_util.go]: I'd say go for it. Please do credit the author and refer to this PR for sure.

Comment (elaron) [reply pkg/volume/rbd/rbd_util.go]: I have send two emails to this PR's author, but get no reply from him. so sad!

**Pull request closed without merging**